File: devices/sources/target_ptr_map.5.c
/*
* @@name:	target_ptr_map.5
* @@type:	C
* @@operation:	compile
* @@expect:	success
* @@version:	omp_5.2
*/
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>

typedef struct {
   int *ptr;
   int buf_size;
} T;

#pragma omp declare mapper(deep_copy: T s) map(s, s.ptr[:s.buf_size])

void do_work(int *ptr, const int size);

int main() 
{
   const int n = 1000;
   const int buf_size = sizeof(int) * n;
   T s = { 0, buf_size };   
   const int dev = omp_get_default_device();
   s.ptr = (int *)malloc(buf_size);
   const int accessible =
      omp_target_is_accessible(s.ptr, s.buf_size, dev);

   #pragma omp metadirective \
      when(user={condition(accessible)}: target) \
      otherwise(target map(mapper(deep_copy),tofrom:s) )
   {
      do_work(s.ptr, n);
   } 

   free(s.ptr);
   return 0;
}

File: devices/sources/target_ptr_map.5.f90
! @@name:	target_ptr_map.5
! @@type:	F-free
! @@operation:	compile
! @@expect:	success
! @@version:	omp_5.2
program main
   use omp_lib

   use, intrinsic :: iso_c_binding, only : c_loc, c_size_t, c_sizeof, c_int
   implicit none
   external :: do_work

   type T
     integer,pointer :: ptr(:)
     integer         :: buf_size
   end type

   !$omp declare mapper(deep_copy: T :: s) map(s, s%ptr(:s%buf_size))

   integer,parameter :: n = 1000
   integer(c_int)    :: dev, accessible
   integer(c_size_t) :: buf_size

   type(T) s

   allocate(s%ptr(n))

   buf_size = c_sizeof(s%ptr(1))*n
   dev = omp_get_default_device()

   accessible = omp_target_is_accessible(c_loc(s%ptr(1)), buf_size, dev)

   !$omp  begin metadirective                              &
   !$omp&       when(user={condition(accessible)}: target) &
   !$omp&       otherwise( target map(mapper(deep_copy),tofrom:s) )

      call do_work(s, n)

   !$omp  end   metadirective

   deallocate(s%ptr)

end program
